go/src/koding/klientctl/commands/cli/cli.go

```diff
@@ -0,0 +1,145 @@
+package cli
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"koding/kites/metrics"
+	"koding/klientctl/config"
+	"koding/klientctl/util"
+
+	"github.com/koding/logging"
+)
+
+// PrintJSON converts provided object to formatted JSON string and writes it to w.
+func PrintJSON(w io.Writer, v interface{}) {
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "\t")
+	enc.Encode(v)
+}
+
+// CLI represents the kd command line client that stores data streams and basic
+// information about kd state.
+type CLI struct {
+	in  io.ReadCloser // input stream.
+	out io.Writer     // output stream.
+	err io.Writer     // error stream.
+
+	m *metrics.Metrics // usage metrics.
+
+	debug bool
+	log   logging.Logger
+}
+
+// NewCLI creates a new CLI client.
+func NewCLI(in io.ReadCloser, out, err, logHandler io.Writer) *CLI {
+	c := &CLI{
+		in:    in,
+		out:   out,
+		err:   err,
+		debug: isDebug(),
+		log:   newLogger(logHandler),
+	}
+
+	if !config.Konfig.DisableMetrics {
+		if m, err := metrics.New("kd"); err != nil {
+			c.Log().Warning("Metrics will not be collected: %v", err)
+		} else {
+			c.m = m
+		}
+	}
+
+	return c
+}
+
+// In returns CLI's input stream. Defaults to standard input when nil.
+func (c *CLI) In() io.Reader {
+	if c.in != nil {
+		return c.in
+	}
+
+	return os.Stdin
+}
+
+// Out returns CLI's output stream. Defaults to standard output when nil.
+func (c *CLI) Out() io.Writer {
+	if c.out != nil {
+		return c.out
+	}
+
+	return os.Stdout
+}
+
+// Err return CLI's error stream. Defaults to standard error when nil.
+func (c *CLI) Err() io.Writer {
+	if c.err != nil {
+		return c.err
+	}
+
+	return os.Stderr
+}
+
+// Log returns CLI's logger. Defaults to discard logger when nil.
+func (c *CLI) Log() logging.Logger {
+	if c.log != nil {
+		return c.log
+	}
+
+	return newLogger(nil)
+}
+
+// Metrics returns metrics client or nil if not enabled.
+func (c *CLI) Metrics() *metrics.Metrics {
```

Comment:
Yes, we can. However, I'm still not sure if this becomes an interface (like in Docker).

Moreover, I wanted to not add a public member to the structure which has only getters(consistency).